### PR TITLE
🚨🚨🚨 Optimize Top P Sampler and fix edge case

### LINKED ISSUE #18976: Top_P sampling samples an extra token when the cum sum of probabilities is exactly equal to top_p
### System Info

- `transformers` version: 4.20.1
- Platform: Linux-5.10.133+-x86_64-with-debian-bullseye-sid
- Python version: 3.7.12
- Huggingface_hub version: 0.8.1
- PyTorch version (GPU?): 1.11.0+cpu (False)
- Tensorflow version (GPU?): 2.6.4 (False)
- Flax version (CPU?/GPU?/TPU?): 0.6.0 (cpu)
- Jax version: 0.3.16
- JaxLib version: 0.3.15

### Who can help?

@patrickvonplaten @Narsil @gante

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Top p sampling samples an extra token when the cumulative sum of probabilities of token is exactly equal to the given top p. E.g., if the input probabilities is `[0.3, 0.1, 0.1, 0.5]` and top_p = `0.8` then only 2 tokens with probability `0.5` and `0.3` should be sampled as their sum would exactly be equal to `0.8`. I believe this is the expected behavior of Top P sampling according to the [definition](https://huggingface.co/docs/transformers/main_classes/text_generation) which states that: 
top_p (float, optional, defaults to 1.0) — If set to float < 1, only the most probable tokens with probabilities that add **up to top_p** or higher are kept for generation. 

I have created a notebook which reproduces this behavior. The notebook also has a proposed implementation which will fix this with an added optimization of not needing to clone tensor and shifting to left or right. https://www.kaggle.com/ekagra/hf-contrib-topp

I have checked locally that the proposed implementation passes the existing [unittest ](https://github.com/huggingface/transformers/blob/f7196f2e63b14e9fbb4ad664e71912aab3b484cf/tests/generation/test_generation_logits_process.py#L162).

### Your contribution
If this makes sense then I would be happy to raise a PR for this.

### PR DESCRIPTION
**Pull request by ekagra-ranjan**
# What does this PR do?

This PR does the following:
1. Fixes #18976
2. Optimizes the Top P sampler Pytorch implementation by removing the need to clone an intermediate tensor and shifting things to right.
3. Add edge case test to PT, TF, FLAX

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@gante @patrickvonplaten

### CODE REVIEW

**Comment by ekagra-ranjan** [comment]
Torch tests are passing since the change was for torch. Checking for TF and flax now.

**Review by gante — COMMENTED** [review]
LGTM, will get my approval when the test change also gets added to TF and FLAX 👍

**Comment by ekagra-ranjan** [comment]
@gante the proposed PT implementation passes the edge case. The existing FLAX implementation also passes the edge case. However, the TF implementation passes the edge case when `use_xla` is True but fails when it is false

**Comment by ekagra-ranjan** [comment]
I was investigating on TF's behavior and found this:

This is the input distribution to the test:
https://github.com/huggingface/transformers/blob/a86acb75ad832fd604a1d5b5e5089f299aae5df4/tests/generation/test_generation_tf_logits_process.py#L190

The above goes to TFTopPLogitsWrapper which takes a cumsum here:
https://github.com/huggingface/transformers/blob/a86acb75ad832fd604a1d5b5e5089f299aae5df4/src/transformers/generation_tf_logits_process.py#L173

This `cumulative_probs` gets different value for `use_xla` as True or False in the unittest. 
1. When `use_xla` is True then cumulative_probs is [[0.5, 0.8,  0.90000004, 1.],
       [0.29999998, 0.59999996, 0.8499999 , 0.99999994]]
2. When `use_xla` is False then cumulative_probs is [[0.5, 0.79999995, 0.9, 1. ],
 [0.3, 0.6, 0.85, 1. ]

This is causing an extra sample to get be sampled when `use_xla` is False as 0.79999995 is < 0.8.

**Comment by gante** [comment]
@ekagra-ranjan we could add an if/else depending on whether `use_xla` is True or not, and set `top_p` to `0.8` or `0.79999995` accordingly. 

However, since this edge case has such low impact in practice, it's okay if we take the simpler path and simply set `top_p` to `0.79999995`. It won't test the edge case with XLA, but I'd say it's good enough.

P.S.: TF's softmax is known to have these minor numerical instabilities.

**Comment by ekagra-ranjan** [comment]
@gante Thank you for your reviews! Edge case test for FLAX and TF have been added and are passing

**Review by gante — APPROVED** [review]
LGTM 👍 Thanks for addressing all the comments!

File: src/transformers/generation_logits_process.py
Comment (patrickvonplaten) [inline src/transformers/generation_logits_process.py:198]: Isn't `<=` always a bit dangerous with float values? I'm not sure we can assure 100% backward compatibility here

Comment (patrickvonplaten) [reply src/transformers/generation_logits_process.py:198]: Slightly worried about that we'll silently break someone's PyTorch generation code that uses `top_p` by default here

Comment (gante) [reply src/transformers/generation_logits_process.py:198]: @patrickvonplaten there is indeed a change at the edge case -- before, if `top_p` was `0.8` and the input was `[0.5, 0.3, 0.1, 0.1]`, the first three tokens would pass this filter, despite the first two summing up to 0.8 (and thus satisfying the top P conditions, according to the original paper and our docstrings).

The behavior in TF and FLAX satisfies the edge case above, while PT does not. In practice, the impact will be negligible (this change filters one additional token when the sum of the logits is exactly `top_p`), although it can change seeded test cases.

Alternatively, we can change our docstrings (and TF+FLAX's implementation) to ignore this edge case :D

Comment (ekagra-ranjan) [reply src/transformers/generation_logits_process.py:198]: @patrickvonplaten  I believe you are referring to the floating point precision in the context of `<=` being dangerous with float value. The Top P sampler intends to pick minimum elements which have cumulative dist >= `top_p`. So either we use the equality while selecting the mask or ignore it and then shift the mask to right/left. 

The proposed PT implementation uses `<=` but it can be implemented in the same manner as TF and FLAX which do not have the equality operator but will need to clone a tensor and shifting values to right/left. This however will not prevent the issue of floating point precision. 

E.g., if we take input as `[0.5, 0.3, 0.1, 0.1]` and `top_p` as `0.8` then according to this:
https://github.com/huggingface/transformers/blob/693ba2cc79ef46e8323b35f791233ce94f433b5d/src/transformers/generation_flax_logits_process.py#L142-L146

the cumsum _could_ be [0.5, 0.79995, 0.1, 0.1] due to floating point precision which will lead to Top P sampler picking 1st three elements instead of 1st two.

Comment (patrickvonplaten) [reply src/transformers/generation_logits_process.py:198]: Thanks for the explanations @gante and @ekagra-ranjan - this makes sense to me!

Given the very high usage of `generate` and `top_p` we need to clearly mark this as a "**breaking behavior bug fix**" with 🚨🚨🚨 in the PR description and also make sure it's mentioned in our release notes (cc @LysandreJik )

But good for merge then for me

**Review by patrickvonplaten — APPROVED** [review]
Let's just make sure users that have their generation pipeline give different results find this PR here as the explanation. It's a backwards breaking bug change for me that however might affect quite some generation pipelines. IMO it's ok to merge with a big warning - wdyt @LysandreJik @sgugger ?

**Review by sgugger — COMMENTED** [review]
I don't have a problem with fixing the behavior to match the documentation.

For the PR before merging, could you?
1. run `make style` to fix the quality issue
2. rebase on main, which should take care of the tests failures we see

**Comment by ekagra-ranjan** [comment]
@sgugger Sure, done.

**Review by LysandreJik — APPROVED** [review]
Looks good to me!

Will make sure this is very visible in the release notes, thank you for the :rotating_light:

**Pull request merged**
